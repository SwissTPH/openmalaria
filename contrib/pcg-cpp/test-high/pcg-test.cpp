/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
