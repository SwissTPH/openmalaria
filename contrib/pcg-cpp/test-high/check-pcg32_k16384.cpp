#define RNG pcg32_k16384
#define TWO_ARG_INIT 1

#include "pcg-test.cpp"

