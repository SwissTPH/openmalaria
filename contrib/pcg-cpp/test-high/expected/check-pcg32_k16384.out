pcg32_k16384:
      -  result:      32-bit unsigned int
      -  period:      2^524352   (* 2^63 streams)
      -  size:        65552 bytes

Round 1:
  32bit: 0xe7cee806 0x5f20781c 0xfe49a493 0x18715647 0x4624a50a 0x7d1c9694
  Again: 0xe7cee806 0x5f20781c 0xfe49a493 0x18715647 0x4624a50a 0x7d1c9694
  Coins: HHTTHTHHTHTTHHTTTTTTTTTTHHHTTTTTTHHHHHHHHHHHTTHTHHHHTTHHHHHHTHTTT
  Rolls: 1 5 5 1 4 1 4 6 1 6 3 1 3 3 4 6 5 4 6 1 6 3 1 2 6 1 2 1 4 2 5 3 1
   -->   rolling dice used 33 random numbers
  Cards: 4h 3c Qd Qh 4d 6d 9d 5s Jd 3d 4c 6c 9h 7h 2d Kc 6s 7d 4s 3s Kh Ah
	 Qc Ac 8s 7c As Tc 3h 5h 8d Ts 2s 9c Td Qs 9s Ad 8h 8c Th Js Kd Jh
	 6h 5c 2c 7s Ks 5d Jc 2h

Round 2:
  32bit: 0x572f5c46 0xcf87184a 0x61374621 0x9ee8ee68 0x04a606ce 0x2ecb1cdf
  Again: 0x572f5c46 0xcf87184a 0x61374621 0x9ee8ee68 0x04a606ce 0x2ecb1cdf
  Coins: THTHTHHTTTHTTHTTTHHHTTHTHHTHTTHHTTHHHHTHTTTHHHTTHHHHHHHHHTTTTHTHH
  Rolls: 6 5 6 1 3 2 4 3 2 2 4 2 1 5 5 2 3 3 4 6 6 2 3 4 1 3 4 6 2 4 5 4 6
   -->   rolling dice used 33 random numbers
  Cards: 5d Th 2d Qs Qd 6s 6d Kh 2c Kd Ad 4s Ah 6c Td Ac 4h 9s 7s Tc 8c Jc
	 Jh 9h 8d 7h 2h Ts 5c Kc Jd 3d 8s 3c 2s Qc 7c 7d Ks 6h 5h 8h 4c Js
	 Qh 4d 3s 9d 9c As 5s 3h

Round 3:
  32bit: 0xc21b1b6b 0xe17a116c 0x21895137 0xc0a7923d 0x389c0b3b 0x4a16d90e
  Again: 0xc21b1b6b 0xe17a116c 0x21895137 0xc0a7923d 0x389c0b3b 0x4a16d90e
  Coins: THTTTHHTTHHTHTTTTHHHTTHHTTHTHTHTTTHHHTHHHHHTHHHHHTHHHHHHHTTTTHTTH
  Rolls: 4 4 2 2 3 2 6 1 3 5 1 3 6 2 6 3 3 4 3 5 4 5 4 4 3 3 2 5 6 6 3 3 2
   -->   rolling dice used 33 random numbers
  Cards: 7s Qd Ah 4c 6d 4d Ac 8d Tc As 2c Jc Qs 6c 3c 9s 5h 4s Kh Kd 2d Jd
	 9d 9c Ks Jh Kc 8c Td 3s 2h Ts 5c 4h 7h 9h Th 7d Js 7c 3h 6s Qc 5s
	 8h 5d 8s 6h Ad 3d Qh 2s

Round 4:
  32bit: 0x72fea21a 0x0a08832b 0x4daa4481 0x15c072de 0x84278d28 0x426515f4
  Again: 0x72fea21a 0x0a08832b 0x4daa4481 0x15c072de 0x84278d28 0x426515f4
  Coins: HHHTTTHHTHHHHHHTTTTHHHHTTTHTTTTTTTTHHHHHHTTHTHHTTTTHTTHHHHHHHTTHH
  Rolls: 4 5 6 3 4 2 2 3 6 6 2 3 2 1 6 5 1 1 6 6 3 4 6 2 2 1 2 6 2 3 3 6 3
   -->   rolling dice used 33 random numbers
  Cards: 2h 6s Tc As 2c Ts 2s 5s 6c 8d 8s 7h 3h Ad 7c 9c Kc 4c 3c 5d 3s Qd
	 Ac 9d Qs Td 5c Th 2d Js 9s 8h 6h Qh Ks 4d Kd Jc 4s Jd Qc 6d 9h 7s
	 7d 8c 5h Kh 3d 4h Ah Jh

Round 5:
  32bit: 0x253cb0a3 0x8c048602 0xe2c21841 0x948f5ed1 0xe460f940 0x606e28a7
  Again: 0x253cb0a3 0x8c048602 0xe2c21841 0x948f5ed1 0xe460f940 0x606e28a7
  Coins: HHTTHHTTTHTHHTTHHHTTHHTHHHHHTHHTHTTTTTHTTHHHTHTHTHTHHTHTTHHHTTTHT
  Rolls: 2 2 4 4 3 1 4 2 2 3 1 4 4 6 4 2 6 4 5 3 4 2 5 5 5 1 1 1 6 6 6 2 1
   -->   rolling dice used 33 random numbers
  Cards: 2h Kh 6h Jc 3d Kd 7h 6s 5h Jh Ks 8d 6d Ad 4c 2s 7c Ah Js 7d 7s 6c
	 Td 3s 8h 5d 4h As 3c Ts Qc 9d 2c Kc Jd 9s 5s 2d 8s 4d 3h Tc Qs 8c
	 9c 4s 5c Qd Qh Th Ac 9h

